aws-ecs/README.md

```diff
@@ -0,0 +1,408 @@
+  http://foo.region.compute.amazonaws.com
+  http://bar.region.compute.amazonaws.com
+  http://baz.region.compute.amazonaws.com
+```
+
+## What did just happen? ##
+
+The automatic `setup.sh` script has:
+
+* Created an ECS cluster named `weave-ecs-demo-cluster`
+* Spawned three hosts (EC2 instances in Amazon's jargon), which are now part of
+  the cluster and are based on Weave's ECS
+  [AMI](https://en.wikipedia.org/wiki/Amazon_Machine_Image).
+* Created a ECS task family definition, which describes the HTTP Server and Data Producer containers.
+* Spawned three tasks, one per host, resulting in an HTTP Server and Data Producer container running in each host.
+
+## Testing the setup ##
+
+The three URLs at the end of `setup.sh`'s output will communicate your browser
+with the HTTP Server containers. Pick one of them (or take the three of them if
+you like) and open them in your browser. This is what you should see:
+
+![httpserver's output](img/httpserver.png)
+
+Pressing reload in your browser will cause HTTP Server to refresh its Data
+Provider address list (generated randomly by `WeaveDNS`) resulting in
+load-balancing between them.
+
+## How does service discovery and load balancing work? ##
+
+Both the HTTP Server and Data Producer containers are very simple (naive if you
+may) to make easy to understand how they work.
+
+They are implemented in a few lines of bash, mainly using Netcat and Dig. 
+
+Container `dataproducer`:
+
+```bash
+while true; do
+  echo 'Hi, this is the data producer in' `hostname -i` | nc -l -p 4540
+done
+```
+
+Container `httpserver`:
+
+```bash
+sleep 7 # Wait for data producers to start
+while true; do
+  # Get the IPs of all the Data Producers
+  DATA_PRODUCER_IPS=`dig +short dataproducer`
+  # Get a message from the first IP in the list
+  DATA_PRODUCER_MESSAGE=`nc ${DATA_PRODUCER_IPS%%$'\n'*} 4540`
+  HTML="<html> <head> <title>Weaveworks Amazon ECS Sample App<\/title> <style>body {margin-top: 40px; background-color: #333;} <\/style> <\/head><body> <div style=color:white;text-align:center> <h1>Data producer addresses:<\/h1> <h2>${DATA_PRODUCER_IPS}<\/h2><h1>Chosen data producer message:<\/h1> <h2>${DATA_PRODUCER_MESSAGE}<\/h2> <\/div>"
+  echo "$HTML" | nc -l -p 80
+done
+```
+
+Note that, for clarity purposes, the sources above have been slightly reformatted.
+
+At launch time, WeaveDNS will register A-records based on the container's name:
+
+* A `dataproducer` A-record for all the Data Producer containers.
+* A `httpserver` A-record for all the HTTP Server container.
+
+The Data Producer waits for requests on TCP port 4540 and responds with a string containing its IP.
+
+The HTTP Server works as follows:
+
+1. Finds out the IP of all the Data Producers by making a DNS query for hostname
+   `dataproducer` (`dig +short dataproducer`).  This works because WeaveDNS
+   keeps A-records for all of them. Also, for load balancing purposes, WeaveDNS
+   will randomize the order of the IPs in every request.
```

Comment:
If you leave in the `dig`, users *will* think that in order to connect to another container they have invoke `dig +short`, parse the output, pick an entry from the result, and then use that as the connection string.

You say that the "HTTP Server and Data Producer containers are very simple (naive if you
+may)", yet the connection establishment is actually considerably more complicated than what a user should be doing.

As I said, by all means use `dig` in the explanation. It has no place in the introductory example, which should be ruthlessly simple, with absolutely nothing extraneous.